go-controller/pkg/ovn/gateway_cleanup.go

```diff
@@ -159,3 +159,96 @@ func staticRouteCleanup(nextHops []net.IP) {
 		}
 	}
 }
+
+// multiJoinSwitchGatewayCleanup removes all the OVN NB gateway logical entities created based
+// on the obsoleted multiple join switch OVN topology. This function is used for upgrading OVN
+// logical topology to single join switch version.
+func multiJoinSwitchGatewayCleanup(nodeName string) error {
+	gatewayRouter := gwRouterPrefix + nodeName
+
+	// Get the gateway router port's IP address (connected to join switch)
+	var nextHops []net.IP
+
+	gwIPAddrs, err := util.GetLRPAddrs(gwRouterToJoinSwitchPrefix + gatewayRouter)
+	if err != nil {
+		return err
+	}
+
+	for _, gwIPAddr := range gwIPAddrs {
+		// Delete logical router policy whose nexthop is the old rtoj- gateway port address
+		stdout, stderr, err := util.RunOVNNbctl("--data=bare", "--no-heading", "--columns=_uuid",
+			"find", "logical_router_policy", fmt.Sprintf("nexthop=%s", gwIPAddr.IP))
+		if err != nil {
+			klog.Errorf("Unable to find LR policy of nexthop: %s for node %s, stderr: %s, err: %v",
+				gwIPAddr.IP, nodeName, stderr, err)
+			continue
+		}
+		if stdout != "" {
+			policyIDs := strings.Fields(stdout)
+			for _, policyID := range policyIDs {
+				_, stderr, err = util.RunOVNNbctl("remove", "logical_router", ovnClusterRouter, "policies", policyID)
+				if err != nil {
+					klog.Errorf("Unable to remove LR policy: %s, stderr: %s, err: %v", policyID, stderr, err)
+				}
+			}
+		}
+		nextHops = append(nextHops, gwIPAddr.IP)
+	}
+	staticRouteCleanup(nextHops)
+
+	// Remove the join switch that connects ovn_cluster_router to gateway router
+	_, stderr, err := util.RunOVNNbctl("--if-exist", "ls-del", "join_"+nodeName)
+	if err != nil {
+		return fmt.Errorf("failed to delete the join logical switch %s, "+
+			"stderr: %q, error: %v", "join_"+nodeName, stderr, err)
+	}
+
+	// Remove the gateway router associated with nodeName
+	_, stderr, err = util.RunOVNNbctl("--if-exist", "lr-del",
+		gatewayRouter)
+	if err != nil {
+		return fmt.Errorf("failed to delete gateway router %s, stderr: %q, "+
+			"error: %v", gatewayRouter, stderr, err)
+	}
+
+	// Remove external switch
+	externalSwitch := externalSwitchPrefix + nodeName
+	_, stderr, err = util.RunOVNNbctl("--if-exist", "ls-del",
+		externalSwitch)
+	if err != nil {
+		return fmt.Errorf("failed to delete external switch %s, stderr: %q, "+
+			"error: %v", externalSwitch, stderr, err)
+	}
+
+	// Remove the patch port on the distributed router that connects to join switch
+	_, stderr, err = util.RunOVNNbctl("--if-exist", "lrp-del", "dtoj-"+nodeName)
+	if err != nil {
+		return fmt.Errorf("failed to delete the patch port dtoj-%s on distributed router "+
+			"stderr: %q, error: %v", nodeName, stderr, err)
+	}
+
+	// If exists, remove the TCP, UDP load-balancers created for north-south traffic for gateway router.
```

Comment:
Is this really required? This is configured on a per GR basis, and we have some sync logic when services come up to look for stale reject ACLs. I think for that we need the load balancers to exist.

Reply:
The upgradeToSingleSwitchOVNTopology() function is used to do two things:

Reply:
This logic here: https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/ovn/service.go#L104

I don't see why for upgrade we need to delete anything other than the join switch and the join switch ports. Those load balancers are also configured on the node logical switches. I think we need to only delete what is required.

Reply:
Ok, I added an argument to the multiJoinSwitchGatewayCleanup() function to differentiate the upgrade case and syncNode() equivalence case. Please take a look and see if it is ok. Thanks

Reply:
Thanks